Create char_array_extents.cpp

willwray opened the pull request:
https://wandbox.org/permlink/5aVvo78zl2w0nYIl
Not the answer you're looking for, and disqualifies itself for using includes.
Also, it can't compile for strings as large as "StringProxy" because the resulting char array type is too large
(even though it is never instantiated)

willwray committed:
Create char_array_extents.cpp

https://wandbox.org/permlink/5aVvo78zl2w0nYIl
Not the answer you're looking for, and disqualifies itself for using includes.
Also, it can't compile for strings as large as "StringProxy" because the resulting char array type is too large
(even though it is never instantiated)

willwray committed:
Fix bug in string proxy operator==

willwray committed:
Create not_so_pretty.cpp

A hack using __PRETTY_FUNCTION__ on recent GCC to extract the name of the type as the proxied string.
Here, the type is an (incomplete) struct declared at point-of-use.
This would be slightly less hacky using function_name() from std::experimental::source_location,
but includes are disallowed.

This solves the puzzle as stated, even if it doesn't port across compilers or compiler versions.
 https://wandbox.org/permlink/pvNgUy6djGQbUKSD

amarmer reviewed (commented):
It cannot be considered as a solution, because it is the same as using hardcoded "StringProxy" string in the function StringProxy. StringProxy is a generic function which reconstructs any C-style string from it's  proxy template parameter, as it is described in the puzzle.

willwray commented:
Which one? I posted two very different methods.
Both encode the string in the type of template argument
(i.e. neither one has the string hardcoded in the function).

**char_array_extents.cpp** encodes the string's characters in the extents of a multi-dimensional char array type, so `char['S']['t']['r']['i']['n']['g']` encodes the string "String". The sizeof an instance of this type would be huge - 83 x 116 x 114 x 105 x 110 x 103 = 1.848137797×10¹⁴ bytes. In theory this should not be a problem as the type is never instantiated, but compilers appear to impose a limit so "StringProxy" encoding is rejected as too large. This method is disqualified according to your 'no includes' rule - decoding the array extents without using `type_traits` and `index_sequence` would be tedious.

**not_so_pretty.cpp** is not pretty - it uses a function 'pretty-print' compiler extension to capture the substring for the name of the function's template parameter. So, the string is not hardcoded in the function, it is passed as the name of type argument. In the posted code I use `struct StringProxy` as the (incomplete) type, to generate a "StringProxy" string proxy, as required by the rules (also, it shows that there is no name collision with the StringProxy template function). Any other name could be used to encode the string of that name.
For instance https://wandbox.org/permlink/FoTSGMZZLD4dUFnj
```
   static_assert(StringProxy<struct String>() == "String", "");
```
This follows the rules as stated; I hereby claim my free beer.

I'd understand if you were to update the rules to say "standard C++ - no compiler extensions".

>StringProxy is a generic function which reconstructs any C-style string from it's proxy template parameter

This does not scan right to me as you cannot 'reconstruct a C-style string' that will compare equal with a string literal. The puzzle wording is 'reconstruct the compile-time string' and, along with the StringProxy name, implies that the solution should return a proxy type that implements operator== as a char-by-char test.

amarmer commented:
My comment was about second method, I misunderstood it (because of compiler extension), please disregard it.
I assume that "Standard C++" implies by default (especially considering that it is in isocpp.org :-).

willwray committed:
Create function_array_extents.cpp

https://wandbox.org/permlink/mJAY6sjHfXQwJV3y

This solution uses function types to represent the string and its proxy comparator.
The string characters are encoded as 1D array extents in a function parameter list:

    The string with chars `c...` is represented as `void(*)(char(&)[c]...)`

Then, this string representation is embedded as the second parameter in:

    'Proxy' function signature `bool( char const*, void(*)(char(&)[c]...) )`

This 'Proxy' function type is passed as the template type parameter to StringProxy.
The StringProxy function unpacks the variadic parameter list using a generic lambda
(effectively a template in disguise - templates are otherwise forbidden in local scope)
with deduction happening as the captureless lambda converts to a function pointer.
An explicit template parameter list `<char... c>` captures the actual chars
(a nice new C++20 feature, currently only implemented in gcc).

The 'Proxy' signature becomes the comparator, implemented by the lambda.

I hereby claim my free beer.

amarmer commented:
Interesting solution!
However, there is a solution which compiles with gcc and c!ang C++17.

ClaasBontus commented:
[https://gcc.godbolt.org/z/3Hcn8H](https://gcc.godbolt.org/z/3Hcn8H)

OK, it has code outside of `PUZZLE`. But did I understand the challenge correctly?

willwray committed:
Create recursive_function_array_extents.cpp

A recursive version of `function_array_extents.cpp`.
Compiles and passes with Clang and GCC -std=c++17 (haven't tested MSVC yet).
gcc https://wandbox.org/permlink/frYkPr1ThmTJtana
clang https://wandbox.org/permlink/PZGssv8wxhyR2L7N

Each string character is represented as the extent of a single array reference parameter
of a function that returns the next 'character' and so on recursively.

This posted version bends the rules by defining a StringProxy specialization,
for the end-of-string type, in order to easily terminate the recursion
(all code is still within StringProxy local scope, so sort of in the spirit of the puzzle).

Here are versions which work without the specialization
but for Clang and GCC separately... 
gcc https://wandbox.org/permlink/KGjkXsJSJirujOvO
clang https://wandbox.org/permlink/SMcCH3XF1pGo2vQV

willwray committed:
Update recursive_function_array_extents.cpp

A tightened-up version of the recursive function solution.
It now compiles on both Clang & GCC -std=c++17,
and the rule-bending specialization  has been removed.

Clang https://wandbox.org/permlink/yRCTkhHG93EPp6t9
GCC https://wandbox.org/permlink/UxWTyE90taZl2Mnk

willwray committed:
Update recursive_function_array_extents.cpp

Remove redundant lambda
